lib/src/main/java/org/kpipe/consumer/OffsetManager.java

```diff
@@ -0,0 +1,574 @@
+    }
+  }
+
+  private OffsetManager(final Builder<K, V> builder) {
+    this.consumer = builder.consumer;
+    this.commitInterval = builder.commitInterval;
+    this.executorService = Executors.newVirtualThreadPerTaskExecutor();
+  }
+
+  /**
+   * Starts the OffsetManager and begins periodic offset commits. This method is idempotent -
+   * calling it multiple times has no effect if the manager is already started.
+   *
+   * @return this instance for method chaining
+   * @throws IllegalStateException if the OffsetManager is already closed
+   */
+  public OffsetManager<K, V> start() {
+    if (state.get() == OffsetState.STOPPED) {
+      throw new IllegalStateException("Cannot restart a stopped OffsetManager");
+    }
+
+    if (state.compareAndSet(OffsetState.CREATED, OffsetState.RUNNING)) {
+      scheduler =
+        Executors.newSingleThreadScheduledExecutor(r ->
+          Thread.ofPlatform().daemon().name("offset-commit-scheduler").unstarted(r)
+        );
+
+      scheduledCommitTask =
+        scheduler.scheduleAtFixedRate(
+          this::commitSafeOffsets,
+          commitInterval.toMillis(),
+          commitInterval.toMillis(),
+          TimeUnit.MILLISECONDS
+        );
+
+      LOGGER.info("OffsetManager started with commit interval of " + commitInterval);
+    }
+
+    return this;
+  }
+
+  /**
+   * Tracks an offset that is about to be processed.
+   *
+   * @param record The consumer record to track
+   */
+  public void trackOffset(final ConsumerRecord<K, V> record) {
+    trackOffset(new TopicPartition(record.topic(), record.partition()), record.offset() + 1);
+  }
+
+  /**
+   * Tracks an offset that is about to be processed.
+   *
+   * @param partition The topic partition the offset belongs to
+   * @param offset The offset to track
+   */
+  public void trackOffset(final TopicPartition partition, final long offset) {
+    if (state.get() == OffsetState.STOPPED) {
+      return;
+    }
+
+    pendingOffsets.computeIfAbsent(partition, k -> new ConcurrentSkipListSet<>()).add(offset);
+
+    // Initialize the next offset to commit if not already set
+    if (!nextOffsetsToCommit.containsKey(partition)) {
+      initializePartitionOffset(partition);
+    }
+  }
+
+  /** Initialize the offset position for a partition. */
+  private void initializePartitionOffset(final TopicPartition partition) {
+    enqueueCommand(
+      () -> {
+        try {
+          nextOffsetsToCommit.put(partition, consumer.position(partition));
+        } catch (final Exception e) {
+          LOGGER.log(WARNING, "Failed to get position for partition " + partition, e);
+        }
+      },
+      "initPosition"
+    );
+  }
+
+  /**
+   * Marks an offset as successfully processed.
+   *
+   * @param record The consumer record that was processed
+   */
+  public void markOffsetProcessed(final ConsumerRecord<K, V> record) {
+    markOffsetProcessed(new TopicPartition(record.topic(), record.partition()), record.offset() + 1);
```

Comment:
i understand that offset to commit is processed offset + 1 - but i would move the +1 logic into the commit block. As it is not really semantically matching the tracking - here the method seems like its tracking offset that was processed - which is not offset +1 in itself.

Reply:
@rkolesnev I have updated the documentation a bit to reflect the different behaviors of 

```
public void trackOffset(final ConsumerRecord<K, V> record)

public void trackOffset(final TopicPartition partition, final long offset)

public void markOffsetProcessed(final ConsumerRecord<K, V> record)

public void markOffsetProcessed(final TopicPartition partition, final long offset) 

```

I also tagged you on the relevant property tests in OffsetManagerTest, let me know what you think